### PR TITLE
 Added functions that assemble the URL

### PR DESCRIPTION
**Pull request by Jose-Perigolo**
Functions used by providers that don't have an SDK.
      
      makeUrl: Assemble the queries according to the url received;
      makeConfig: Assemble the request configs;
      getJSON: Assemble get request;
      postJSON,: Assemble post requests;
      deleteJSON: Assemble delete requests;

### CODE REVIEW

File: src/provider.ts [hunk 1]
```diff
@@ -138,10 +144,102 @@ function provider(this: any, options: ProviderOptions) {
...
+        url +=
+        "?" +
+        Object.entries(q)
+        .reduce((u: any, kv: any) => (u.append(kv[0], kv[1]), u), new URLSearchParams())
+        .toString();
+      }
+    }
+  
+    return url;
+  };
```
Comment (rjrodger) [inline src/provider.ts]: no semis in seneca code

File: src/provider.ts [hunk 2]
```diff
@@ -25,6 +25,12 @@ type ProviderOptions = {
   provider: Record<string, Provider>
 }
 
+type UtilitiesProviderOptions = {
```
Comment (rjrodger) [inline src/provider.ts]: rename to `ProviderUtilityOptions` - this orders the concepts in importance

File: src/provider.ts [hunk 3]
```diff
@@ -138,10 +144,102 @@ function provider(this: any, options: ProviderOptions) {
...
+      }
+      err.status = res.status;
+      throw err;
+    }
+  };
 
   return {
     exports: {
-      entityBuilder
+      entityBuilder,
```
Comment (rjrodger) [inline src/provider.ts]: OK! So let's refactor a little :)

We want to make the API a little easier to use. Most of these utility functions are parameterised by options, so let's instead have one exported function:

`makeUtils(utilopts: ProviderUtilityOptions)`

this returns an object:
```
{
      makeUrl,
      makeConfig,
      getJSON,
      postJSON,
      deleteJSON
}
```

Now we can remove the `options` parameter from each of these utilities.

Thus:

```
getJSON(makeUrl("checks", msg.q, options), options, makeConfig(this))
```

becomes:

```
getJSON(makeUrl("checks", msg.q), makeConfig(this))
```

which is much cleaner

File: src/provider.ts [hunk 4]
```diff
@@ -138,10 +144,102 @@ function provider(this: any, options: ProviderOptions) {
...
+  }
+  
+  async function getJSON (url: string, options: UtilitiesProviderOptions, config?: any) {
+    const res = await options.fetch(url, config);
+  
+    if (200 == res.status) {
+      let json: any = await res.json();
+      return json;
+    } else {
+      let err: any = new Error("ChecklyhqProvider " + res.status);
```
Comment (rjrodger) [inline src/provider.ts]: just `Provider`, as this has to work for all providers

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: src/provider.ts (hunk 4 above)
Comment (Jose-Perigolo) [reply src/provider.ts]: Ok, forgot to change

File: src/provider.ts (hunk 3 above)
Comment (Jose-Perigolo) [reply src/provider.ts]: Got it! Indeed, much cleaner.

File: src/provider.ts (hunk 2 above)
Comment (Jose-Perigolo) [reply src/provider.ts]: Ok

File: src/provider.ts [hunk 5]
```diff
@@ -139,9 +145,115 @@ function provider(this: any, options: ProviderOptions) {
   }
 
 
+
+  function makeUtils(utilopts: ProviderUtilityOptions, config?: any) {
+    function makeUrl(suffix: string, q: any) {
```
Comment (rjrodger) [inline src/provider.ts:150]: place the function definitions outside the main provider function
this way we ensure there no accidental variable closures, which could introduce hidden state

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: src/provider.ts (hunk 5 above)
Comment (Jose-Perigolo) [reply src/provider.ts:150]: So, the provider's function is being exported by default in the `provider.ts` file, maybe it's a silly question, but how would I export the `makeUtils` function if it's outside the `provider` ? And if I were to call `makeUtils` within the provider, how would I pass the utilopts and config parameters?

```
    export default provider
    if ('undefined' !== typeof (module)) {
      module.exports = provider
    }
```

**Pull request merged**